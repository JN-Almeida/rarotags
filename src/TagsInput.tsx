import React, { useEffect } from "react";
import Chip from "@material-ui/core/Chip";
import TextField from "@material-ui/core/TextField";

export default function TagsInput(props) {
  const { onSelectTags, tags, ...other } = props;
  const [ selectedItem, setSelectedItem ] = React.useState([]);
  useEffect(() => {
    setSelectedItem(tags);
  }, [tags]);

  const handleDelete = (item) => () => {
    return item;
  };

  return (
    <TextField
      InputProps={{
        startAdornment: selectedItem.map((item) => (
          <Chip
            key={item}
            tabIndex={-1}
            label={item}
            onDelete={handleDelete(item)}
          />
        ))
      }}
      {...other}
    />
  );
}
TagsInput.defaultProps = {
  tags: []
};
