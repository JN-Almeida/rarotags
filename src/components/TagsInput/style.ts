import styled from 'styled-components';

export const Container = styled.div`

    .tagInput {
        &> div{
            flex-wrap: wrap;
            padding: 10px;
            gap: 5px;
            input{
                flex: 1;
                padding: 5px;
            }
        }
    }

    button{
        background-color: #106fdc;
        border-radius: 5px;
        color: #ffffff;
        transition: .5s ease-in-out;
        padding: 10px 60px;
        width: 100%;
        margin: 30px 0;

        &:hover{
            background-color: #00499c;
        }
    }
`;
