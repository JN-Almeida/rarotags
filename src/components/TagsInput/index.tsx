import React, { KeyboardEvent, useEffect, useState } from 'react'
import Chip from '@material-ui/core/Chip'
import TextField from '@material-ui/core/TextField'

import { Container } from './style';

const emailValidation = /^(([^<>()[\]\\.,;:\s@\"]+(\.[^<>()[\]\\.,;:\s@\"]+)*)|(\".+\"))@((\[[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}\])|(([a-zA-Z\-0-9]+\.)+[a-zA-Z]{2,}))$/;

export default function TagsInput(props) {
  const { onSelectTags, tags, setTags, ...other } = props
  const [selectedItem, setSelectedItem] = useState<string[]>([])
  const [inputValue, setInputValue] = useState<string>('')
  useEffect(() => {
    setSelectedItem(tags)
  }, [tags])

  const handleKeyUp = (event: KeyboardEvent) => {
    const target = event.target as HTMLInputElement;
    const value = target.value.trim()

    if ((event.key === 'Enter' || event.key === 'Tab') && value !== '') {
      event.preventDefault()
      const allEmails = value.split(';').filter((mail: string) => {
        const isSame = selectedItem.some((x) => x === mail)
        return emailValidation.test(mail) && !isSame
      })
      setSelectedItem([...selectedItem, ...allEmails])
      setInputValue('')
    }

    if (selectedItem.length && !value.length && event.key === 'Backspace') {
      setSelectedItem(selectedItem.slice(0, selectedItem.length - 1))
    }
  }

  const handleDelete = (item) => () => {
    const newSelectedItem = [...selectedItem]
    newSelectedItem.splice(newSelectedItem.indexOf(item), 1)
    setSelectedItem(newSelectedItem)
  }

  return (
    <Container>
      <TextField
        InputProps={{
          startAdornment: selectedItem.map((item) => (
            <Chip key={item} tabIndex={-1} label={item} onDelete={handleDelete(item)} />
          )),
        }}
        onKeyDown={handleKeyUp}
        value={inputValue}
        onChange={({ target }) => setInputValue(target.value)}
        {...other}
      />

      <button type="button" onClick={() => setTags(selectedItem)}>
        enviar
      </button>
    </Container>
  )
}
TagsInput.defaultProps = {
  tags: [],
}
