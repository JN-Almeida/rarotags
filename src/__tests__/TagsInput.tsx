import React from 'react'
import TagsInput from '@/components/TagsInput'
import { cleanup, render } from '@testing-library/react'

describe('TagsInput Component', () => {
  afterEach(jest.clearAllMocks)
  afterEach(cleanup)

  it('cria o snapshot do component', () => {
    const container = render(<TagsInput />)
    expect(container.asFragment()).toMatchSnapshot()
  })

  it('deve renderizar as tags enviadas por atributos', () => {
    const emails = ['contato@rarolabs.com.br', 'nao-responda@rarolabs.com.br']

    const { debug } = render(<TagsInput tags={emails} />)

    debug()
  })

  it('deve renderizar tags quando preencher o input e pressionar enter', () => {
    // teste não implementado.
  })

  it('deve renderizar tags quando preencher o input e pressionar tab', () => {
    // teste não implementado.
  })

  it('deve deletar a útima tag criada ao pressionar o botão de backspace', async () => {
    // teste não implementado.
  })
})
