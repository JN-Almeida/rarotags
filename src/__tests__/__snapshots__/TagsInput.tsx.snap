// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TagsInput Component cria o snapshot do component 1`] = `
<DocumentFragment>
  <div
    class="MuiFormControl-root MuiTextField-root"
  >
    <div
      class="MuiInputBase-root MuiInput-root MuiInput-underline MuiInputBase-formControl MuiInput-formControl MuiInputBase-adornedStart"
    >
      <input
        aria-invalid="false"
        class="MuiInputBase-input MuiInput-input MuiInputBase-inputAdornedStart"
        type="text"
        value=""
      />
    </div>
  </div>
</DocumentFragment>
`;
